docs/general/configuring.md
# Discord Economy Super

[![Downloads](https://img.shields.io/npm/dt/discord-economy-super?style=for-the-badge)](https://www.npmjs.com/package/discord-economy-super)
[![Stable Version](https://img.shields.io/npm/v/discord-economy-super?style=for-the-badge)](https://www.npmjs.com/package/discord-economy-super)

<b>Discord Economy Super</b> - Easy and customizable economy module for your [Discord Bot](https://discord.js.org/#/).

## Introduction
You know that the module has a lot of different settings and you can set them up like you want. In this article, we will go through everything about configuring the Economy system and how to set it up for different guilds!

## Available Options
Full list of available Economy options is available [here](https://des-docs.js.org/#/docs/main/1.7.5/typedef/EconomyConfiguration).

Please note: 
1. For JSON version, all the options for are **optional** 
2. For MongoDB version, `connection` object is required to be specified in the configuration for MongoDB version. [Learn more about connecting MongoDB to Economy](https://des-docs.js.org/#/docs/main/1.7.5/general/migrating-to-mongo)

## Passing Configuration to the Module
There are 2 ways of passing the configuration object to the module:
- Specifying the config in the constructor:

```js
new Economy({
    // configuration goes here...
})
```
- Or making the `economy.config.js` or `economy.config.ts` file **(depends on which language your project is using - JavaScript or TypeScript)** in root of your project directory *(that's why specifying the config in constructor in MongoDB version is optional even though the `connection` property is required)*:

```js
/* ./economy.config.js */

/**
 * put `discord-economy-super/mongodb/EconomyItems.d.ts` if using MongoDB version of Economy
 * @type {import('discord-economy-super/EconomyItems.d.ts').EconomyConfiguration} 
 */
const economyConfig = {
    // configuration goes here...
}

module.exports = economyConfig
```

```ts
/* ./economy.config.ts */

// put `discord-economy-super/mongodb/EconomyItems` if using MongoDB version of Economy
import { EconomyConfiguration } from 'discord-economy-super/EconomyItems' 

const economyConfig: EconomyConfiguration {
    // configuration goes here...
}

export default economyConfig
```
If using the config file, the module will use them **first**. 

**Don't specify the options in the constructor if using the file to avoid confusion!**

## Daily, Work & Weekly Rewards Configuration
`dailyAmount`, `workAmount` and `weeklyAmount` setting are accepting the **number** or an **array** of 2 numbers - `[min, max]`.

Passing the `[min, max]` array allows to pick the random number in range between `min` and `max` to add on the user's balance when claiming rewards!

Use a *single number* for static reward amount values.

```js
dailyAmount: [100, 500] // when claiming the daily reward, a random number between 100 and 500 will be picked and added to the user' balance
workAmount: [10, 50] // the same as `dailyAmount` but the range is between 10 and 50
weeklyAmount: 1000 // static value - 1000 will be added to the user's balance on claiming weekly reward, passing the [min, max] array is still allowed.
```

## Configuring Economy for Different Guilds
To edit the configuration keys for a specific guild, you may use the [SettingsManager](https://des-docs.js.org/#/docs/main/1.7.5/class/SettingsManager) for this:
```js
eco.settings.set('key', value, 'guildID')
```
Once the key is edited, **the value from settings** will be used on specified server for the key.
For example, let's change the `dailyAmount` setting to `500` (default is `100`) for guild with ID `'123'`:
```js
eco.settings.set('dailyAmount', 500, '123')
```
From now, everyone on guild with ID `'123'` will receive **500** coins on dailies instead of **100**. This change will affect only our guild with ID `'123'`.

## Full Configuration Example
Here's the **full** example of Economy configuration object and how everything should look like.

This example will contain default values for each setting. 

Note that all settings (except `connection` for MongoDB version) are **optional** and you won't have to specify them all!

```js
const economyConfigExample = {
    storagePath: './storage.json', // specify if using JSON version (optional)
    updateCountdown: 1000, // specify if using JSON version (optional)
    checkStorage: true, // specify if using JSON version (optional)

    connection: { // specify if using MongoDB version (required)
        connectionURI: '...', // mongodb connection URI
        collectionName: 'database', // specify if using MongoDB version (optional)
        dbName: 'db', // specify if using MongoDB version (optional)
        mongoClientProperties: {} // specify if using MongoDB version (optional)
    },

    dailyAmount: 100, // [min, max] array allowed
    workAmount: [10, 50], // [min, max] array allowed
    weeklyAmount: 1000, // [min, max] array allowed
    dailyCooldown: 86400000,
    workCooldown: 3600000,
    weeklyCooldown: 604800000,
    sellingItemPercent: 75,
    savePurchasesHistory: true,
    deprecationWarnings: true,
    dateLocale: 'en',
    subtractOnBuy: true,

    updater: {
        checkUpdates: true,
        upToDateMessage: false
    },

    errorHandler: {
        handleErrors: true,
        attempts: 5,
        time: 5000
    },

    optionsChecker: {
        ignoreInvalidTypes: false,
        ignoreUnspecifiedOptions: true,
        ignoreInvalidOptions: false,
        showProblems: true,
        sendLog: true,
        sendSuccessLog: false
    },

    debug: false
}
```

## ❗ | Useful Links

<ul>
<li><b><a href = 'https://www.npmjs.com/package/discord-economy-super'>NPM</a></b></li>
<li><b><a href = 'https://github.com/shadowplay1/discord-economy-super'>GitHub</a></b></li>
<li><b><a href = 'https://github.com/shadowplay1/discord-economy-super/tree/main/examples'>Bot Examples</a></b></li>
<li><b><a href = 'https://discord.gg/4pWKq8vUnb'>Discord Server</a></b></li>
</ul>
<b>If you found a bug, have any questions or need help, join the <a href = 'https://discord.gg/4pWKq8vUnb'>Support Server</a>.</b>
<br>
<b>Module Created by ShadowPlay.</b>

# ❤️ Thanks for choosing Discord Economy Super ❤️


general/changelog.md
# Discord Economy Super

[![Downloads](https://img.shields.io/npm/dt/discord-economy-super?style=for-the-badge)](https://www.npmjs.com/package/discord-economy-super)
[![Stable Version](https://img.shields.io/npm/v/discord-economy-super?style=for-the-badge)](https://www.npmjs.com/package/discord-economy-super)

<b>Discord Economy Super</b> - Easy and customizable economy framework for your [Discord Bot](https://discord.js.org/#/).

## ⏰ | Changelog
<b>1.0.1</b>
<ul>
  <li><b>The first version of the module: added a basic methods: 'Economy.fetch()', 'Economy.set()', 'Economy.add()', 'Economy.subtract()', 'Economy.daily()', 'Economy.work()', 'Economy.getDailyCooldown()', 'Economy.getWorkCooldown()', 'Economy.all()' and 'Economy.leaderboard()'. Also added an EconomyError class, but you couldn't use it yourself.</b></li>
</ul>
<b>1.0.2</b>
<ul>
  <li><b>Edited README.md</b></li>
</ul>
<b>1.0.3</b>
<ul>
  <li><b>Fixed bugs.</b></li>
</ul>
<b>1.0.4</b>
<ul>
  <li><b>Fixed bugs.</b></li>
</ul>
<b>1.0.5</b>
<ul>
  <li><b>Fixed bugs.</b></li>
</ul>
<b>1.0.6</b>
<ul>
  <li><b>Edited README.md</b></li>
  <li><b>Fixed bugs.</b></li>
  <li><b>Code optimization</b></li>
  <li><b>Now you can create a shop on your Discord server using Economy.shop methods. They are listed above.</b></li>
  <li><b>Added an 'EconomyError' class property.</b></li>
  <li><b>Added a 'dateLocale' property for options object.</b></li>
</ul>
<b>1.0.7</b>
<ul>
  <li><b>Fixed bugs</b></li>
</ul>
<b>1.0.8</b>
<ul>
<li><b>Edted README.md</b></li>
<li><b>Fixed bugs.</b></li>
<li><b>Code optimization.</b></li>
<li><b>Now this module is including Events. They are listed above.</b></li>
</ul>
<b>1.1.4</b>
<ul>
<li><b>Edted README.md</b></li>
<li><b>Fixed bugs.</b></li>
<li><b>Code optimization.</b></li>
<li><b>Added an update checking system for this module.</b></li>
<li><b>Added a 'checkUpdates' property for options object.</b></li>
</ul>
<b>1.1.5</b>
<ul>
<li><b>Edted README.md</b></li>
<li><b>Fixed bugs.</b></li>
<li><b>Code optimization.</b></li>
<li><b>Updated the examples.</b></li>
<li><b>Fixed the types.</b></li>
<li><b>Added a Table of Contents in README.md</b></li>
<li><b>If you have an earlier Node.js version than v14, you will receive an error.</b></li>
<li><b>Now you can configure the update checking system settings using 'options.updater' properties in constructor options. The 'options.checkUpdates' property was removed in favor of 'options.updater' configuration object. Use 'options.updater.checkUpdates' instead.</b></li>
<li><b>Added an error handler that will work on startup. You can configure it using 'options.errorHandler' properties in constructor options.</b></li>
<li><b>Option properties 'description, 'mesasge' and 'maxAmount', in 'Economy.shop.addItem' method are optional now.</b></li>
<li><b>Now everyone will have a Bank Balance. This is another type of money, you can use it whatever you want. :) | The methods are almost the same: 'Economy.bankFetch()', 'Economy.bankSet()', 'Economy.bankAdd()', 'Economy.bankSubtract()' and 'Economy.bankLeaderboard()'.</b></li>
<li><b>Added a new events for Bank Balance: 'bankSet', 'bankAdd' and 'bankSubtract'.</b></li>
<li><b>Now every balance and bank-balance event object is including 'type' and 'balance' properties. The first property is type of the operation ('set', 'add', 'subtract', 'bankSet', 'bankAdd', 'bankSubtract') and the second is user's new balance after the operation was completed successfully.</b></li>
<li><b>Now you can manually check the module updates using the new 'Economy.checkUpdates()' method.</b></li>
<li><b>Added an additional 'Economy.ready' and 'Economy.errored' properties to check the module state.</b></li>
<li><b>Now you can add a role to users on your Discord server. Click <a href="https://des-docs.tk/#/docs/main/stable/class/ShopManager">here</a> for more info.</b></li>
<li><b>Added a new 'shopItemBuy' and 'shopItemUse' events.</b></li>
<li><b>Now if user try to buy an item when he reached the max amount of item in his inventory, the method will return a 'max' string.</b></li>
<li><b>Now this module has a Support Server. Click <b><a href = "https://discord.gg/4pWKq8vUnb">here</a></b> to join!</b></li>
</ul>
<b>1.1.6</b>
<ul>
<li><b>Code optimization.</b></li>
<li><b>Fixed bug with "The module is not ready to work" error on startup.</b></li>
<li><b>Removed an accidentaly added "discord.js" module from dependencies.</b></li>
</ul>
<b>1.1.7</b>
<ul>
<li><b>Fixed bugs.</b></li>
<li><b>Code optimization.</b></li>
<li><b>TypeScript support is finally here! Created a type defenitions for this module.</b></li>
</ul>
<b>1.1.8</b>
<ul>
<li><b>Fixed bugs.</b></li>
<li><b>Code optimization.</b></li>
<li><b>Now you can disable checking a storage file using the 'options.checkStorage' option.</b></li>
<li><b>Now you can kill the Economy instance using the 'Economy.kill()' method.</b></li>
<li><b>'Economy.init()' method is not private anymore due to 'Economy.kill()' method.</b></li>
<li><b>Added an 'interval' property that displays the database checking interval.</b></li>
<li><b>Now you can clear the entire database using the 'Economy.clearStorage()' method.</b></li>
<li><b>Now this module is using errors that in 'Economy.errors' property and in './src/errors.js' file.</b></li>
<li><b>Now you can clear any cooldown using the 'Economy.clearDailyCooldown', 'Economy.clearWorkCooldown' and ''Economy.clearWeeklyCooldown' methods.</b></li>
<li><b>Added a test for basic Economy methods. You can run it by using these commands:

```console
cd node_modules/discord-economy-super
npm test
```
The test will look like this:

![Test](https://cdn.discordapp.com/attachments/764192017542283325/838226957166313472/Screenshot_4.png)
</b></li>
</ul>
<b>1.1.9</b>
<ul>
<li><b>Fixed bugs.</b></li>
<li><b>Code optimization.</b></li>
<li><b>Fixed bugs.</b></li>

src/classes/guild/Settings.js
const DatabaseManager = require('../../managers/DatabaseManager')
const SettingsManager = require('../../managers/SettingsManager')

/**
 * Guild Settings.
 */
class Settings {

    /**
     * Guild settings class.
     * @param {string} guildID Guild ID.
     * @param {EconomyConfiguration} options Economy configuration.
     */
    constructor(guildID, options) {

        /**
         * Guild ID.
         * @type {string}
         * @private
         */
        this.guildID = guildID

        /**
         * Settings Manager.
         * @type {SettingsManager}
         * @private
         */
        this._settings = new SettingsManager(options, new DatabaseManager(options))
    }

    /**
     * Fetches the server's settings object.
     * @returns {SettingsTypes} The server settings object.
     */
    all() {
        return this._settings.all(this.guildID)
    }

    /**
     * Gets the specified setting from the database.
     *
     * Note: If the server don't have any setting specified,
     * the module will take the values from the
     * configuration or default configuration.
     *
     * @param {Settings} key The setting to fetch.
     * @returns {any} The setting from the database.
     */
    get(key) {
        return this._settings.get(key, this.guildID)
    }

    /**
     * Changes the specified setting.
     *
     * Note: If the server don't have any setting specified,
     * the module will take the values from the
     * specified configuration or default configuration.
     *
     * @param {Settings} key The setting to change.
     * @param {any} value The value to set.
     * @returns {SettingsTypes} The server settings object.
     */
    set(key, value) {
        return this._settings.set(key, value, this.guildID)
    }

    /**
     * Removes the specified setting.
     *
     * Note: If the server don't have any setting specified,
     * the module will take the values from the
     * specified configuration or default configuration.
     *
     * @param {Settings} key The setting to remove.
     * @returns {SettingsTypes} The server settings object.
     */
    remove(key) {
        return this._settings.remove(key, this.guildID)
    }

    /**
     * Resets all the settings to setting that are in configuration.
     * @returns {SettingsTypes} The server settings object.
     */
    reset() {
        return this._settings.reset(this.guildID)
    }
}


/**
 * @typedef {object} EconomyConfiguration Default Economy configuration.
 * @property {string} [storagePath='./storage.json'] Full path to a JSON file. Default: './storage.json'
 * @property {boolean} [checkStorage=true] Checks the if database file exists and if it has errors. Default: true
 * @property {number} [dailyCooldown=86400000]
 * Cooldown for Daily Command (in ms). Default: 24 hours (60000 * 60 * 24 ms)
 *
 * @property {number} [workCooldown=3600000] Cooldown for Work Command (in ms). Default: 1 hour (60000 * 60 ms)
 * @property {number | number[]} [dailyAmount=100] Amount of money for Daily Command. Default: 100.
 * @property {number} [weeklyCooldown=604800000]
 * Cooldown for Weekly Command (in ms). Default: 7 days (60000 * 60 * 24 * 7 ms)
 *
 * @property {number | number[]} [weeklyAmount=100] Amount of money for Weekly Command. Default: 1000.
 * @property {number | number[]} [workAmount=[10, 50]] Amount of money for Work Command. Default: [10, 50].
 * @property {boolean} [subtractOnBuy=true]
 * If true, when someone buys the item, their balance will subtract by item price. Default: false
 *
 * @property {number} [sellingItemPercent=75]
 * Percent of the item's price it will be sold for. Default: 75.
 *
 * @property {boolean} [deprecationWarnings=true]
 * If true, the deprecation warnings will be sent in the console. Default: true.
 *
 * @property {boolean} [savePurchasesHistory=true] If true, the module will save all the purchases history.
 *
 * @property {number} [updateCountdown=1000] Checks for if storage file exists in specified time (in ms). Default: 1000.
 * @property {string} [dateLocale='en'] The region (example: 'ru'; 'en') to format the date and time. Default: 'en'.
 * @property {UpdaterOptions} [updater=UpdaterOptions] Update checker configuration.
 * @property {ErrorHandlerConfiguration} [errorHandler=ErrorHandlerConfiguration] Error handler configuration.

 * @property {CheckerConfiguration} [optionsChecker=CheckerConfiguration]
 * Configuration for an 'Economy.utils.checkConfiguration' method.
 * @property {boolean} [debug=false] Enables or disables the debug mode.
*/

/**
 * @typedef {object} SettingsTypes Settings object.
 * @property {number | number[]} dailyAmount Amount of money for Daily Command. Default: 100.
 * @property {number} dailyCooldown Cooldown for Daily Command (in ms). Default: 24 hours (60000 * 60 * 24 ms)
 *
 * @property {number | number[]} workAmount Amount of money for Work Command. Default: [10, 50].
 * @property {number} workCooldown Cooldown for Work Command (in ms). Default: 1 hour (60000 * 60 ms)
 *
 * @property {number | number[]} weeklyAmount Amount of money for Weekly Command. Default: 1000.
 * @property {number} weeklyCooldown Cooldown for Weekly Command (in ms). Default: 7 days (60000 * 60 * 24 * 7 ms)
 *
 * @property {string} dateLocale The region (example: 'ru' or 'en') to format the date and time. Default: 'en'
 * @property {boolean} subtractOnBuy
 * If true, when someone buys the item, their balance will subtract by item price. Default: false.
 *
 * @property {number} sellingItemPercent Percent of the item's price it will be sold for. Default: 75.
 */

/**
 * @typedef {'dailyAmount' | 'dailyCooldown' |
 * 'workAmount' | 'workCooldown' |
 * 'weeklyAmount' | 'weeklyCooldown' |
 * 'dateLocale' | 'subtractOnBuy' |
 * 'sellingItemPercent' | 'savePurchasesHistory'} Settings
 */

/**
 * Guild Settings.
 * @type {Settings}
 */
module.exports = Settings
